core/krylovTimeEvolver.h

```diff
@@ -1,96 +1,101 @@
+    */
     int statusCode;
+    std::vector<std::unique_ptr<krylovBasicObservable>> observableList;
+    std::unique_ptr<smatrix> hamiltonianMatrix;
 
-/* Status code has the following meaning: 
-1-digit codes mean succes: 0 (everything in order, nothing special happened), 1 (lucky breakdown), 2 (computed analytic error is smaller than estimate of numerical error, which in turn is bigger than requested error; so desired error bound is probably respected) 
-more than 1 digit means failure: 10 (computation of error may be  spoiled due to numerical roundoff), 11 (requested tolerance seems unreachable because of roundoff errors), 20 (desired accuracy of numerical integral could not be achieved), 30 (norm of vector deviates significantly from 1), 100 (multiple of these errors)
- */
-    krylovReturn(unsigned int nbObservables, unsigned int Hsize, unsigned int nbSamples, int status)
-    {
-        err = 0; n_steps = 0; dim = Hsize; nSamples = nbSamples; statusCode = status;
-        if(nbObservables == 0 && (nSamples * Hsize * sizeof(std::complex<double>) > std::pow(2.,34.)))
-        {
-            std::cerr << "Requested output would be too large" << std::endl;
-            exit(1);
-        }
-        evolvedState = new std::complex<double>[Hsize];
-        if (nbSamples > 0)
-        {
-            if (nbObservables == 0)
-                sampling = new matrix(Hsize, nbSamples);
-            else {
-                sampling = new matrix(nbObservables, nbSamples);
-            }
-        }
-        else
-            sampling = nullptr;
-    }
-
-	~krylovReturn()
-	{
-            delete[] evolvedState;
-            if (nSamples > 0)
-				delete sampling;
-	}
+    krylovReturn(unsigned int Hsize, int status);
+    ~krylovReturn();
 };
 
-
+/**
+* Main class responsible for computing the actual time evolution.
+*/
 class krylovTimeEvolver
 {
 public:
-    krylovTimeEvolver(double t, size_t Hsize, std::complex<double>* v, double samplingStep, double tol, int mm, smatrix** observables, int nbObservables, smatrix* Ham, std::complex<double> expFactor, bool checkNorm= true, bool fastIntegration = false);
-    krylovTimeEvolver(double t, size_t Hsize, std::complex<double>* v, double samplingStep, double tol, int mm, std::vector<std::unique_ptr<krylovBasicObservable>>  observables, smatrix* Ham, std::complex<double> expFactor, bool checkNorm = true, bool fastIntegration = false, bool progressBar = false);
+    krylovTimeEvolver(double t, std::complex<double>* v, double samplingStep, double tol, int mm, std::vector<std::unique_ptr<krylovBasicObservable>>  observables, std::unique_ptr<smatrix> HamIn, double expFactor = 1.0, bool fastIntegration = false, bool progressBar = true);
```

Comment:
You are right. I kept the order of arguments for legacy reasons but since we changed the datatypes anyway this does not make much sense. 
However, I am not a huge fan of setting so many standard values there. I don't know if it is even possible since calling it (with all standard values set) would be identical to the short one. There would be no way for the compiler to tell which constructor is meant. On top of that, getting the order of importance right would be impossible. The only benefit it would give is that "standard" values would be apparent from the constructor signature. 
I would prefer to mention standard values in the constructor of the short version (or/also somewhere else). 

In summary, I see 2 options. 
1) Kill the second short constructor and set default values for all the values the short constructor does it. 
2a) Keep the long one with an intermediate number of default values
2b) Remove default values in the constructor altogether and mention standard values. (This is the version I would prefer. I spend quite some time in finding apparent errors caused by me messing up the order of default parameters....)

Reply:
Sounds good, let's do 2b then. Still with reordering of arguments, right?

Reply:
cd75537